firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteIndexManager.java

```diff
@@ -87,6 +108,259 @@ public void addFieldIndex(FieldIndex index) {
         true);
   }
 
+  @Override
+  public void addIndexEntries(Document document) {
+    ResourcePath documentPath = document.getKey().getPath();
+    String collectionGroup = documentPath.getSegment(documentPath.length() - 2);
+    db.query(
+            "SELECT index_id, index_proto FROM index_configuration WHERE collection_group = ? AND active = 1")
+        .binding(collectionGroup)
+        .forEach(
+            row -> {
+              try {
+                int indexId = row.getInt(0);
+                FieldIndex fieldIndex =
+                    serializer.decodeFieldIndex(
+                        collectionGroup, row.getInt(0), Index.parseFrom(row.getBlob(1)));
+
+                List<Value> values = extractFieldValue(document, fieldIndex);
+                if (values == null) return;
+
+                if (Logger.isDebugEnabled()) {
+                  Logger.warn(
+                      TAG,
+                      "Adding index values for document '%s' to index '%s'",
+                      documentPath,
+                      fieldIndex);
+                }
+
+                List<byte[]> encodeValues = encodeDocumentValues(fieldIndex, values);
+                for (byte[] encoded : encodeValues) {
+                  // TODO(indexing): Handle different values for different users
+                  db.execute(
+                      "INSERT OR IGNORE INTO index_entries ("
+                          + "index_id, "
+                          + "index_value, "
+                          + "document_name) VALUES(?, ?, ?)",
+                      indexId,
+                      encoded,
+                      documentPath.canonicalString());
+                }
+              } catch (InvalidProtocolBufferException e) {
+                throw fail("Invalid index: " + e);
+              }
+            });
+  }
+
+  /**
+   * Returns the list of values for all fields indexed by the provided field index. Returns {@code
+   * null} if one or more of the fields are not set.
+   */
+  @Nullable
+  private List<Value> extractFieldValue(Document document, FieldIndex fieldIndex) {
+    List<Value> values = new ArrayList<>();
+    for (FieldIndex.Segment segment : fieldIndex) {
+      Value field = document.getField(segment.getFieldPath());
+      if (field == null) {
```

Comment:
If not already, we should have a test where a field is set to null explicitly.

Reply:
I cannot find the new test for this one.

Reply:
`addsDocuments()` is meant to test it. It adds a document that does not have the field being indexed. Note that a field that is set to `null` is being indexed (and its value is `Value.setNullValue()`).